lib/toucan.js

```diff
@@ -63,73 +64,81 @@ async function prepareToucanEnv(admin, broker) {
-	console.log(`Deploying a TCO2 for vintage ${vintageId}...`);
-	await tco2Factory.deployFromVintage(vintageId);
-	const tco2Array = await tco2Factory.getContracts();
-	const tco2 = await ethers.getContractAt("ToucanCarbonOffsets", tco2Array[0]);
-	console.log(`Deployed TCO2 ${tco2.address}`);
-
 	// Admin is able to verify
 	const role = await cBatches.VERIFIER_ROLE();
 	await cBatches.grantRole(role, admin.address);
 
-	// Mint an empty batch to start the bridging process for the broker
-	console.log("Starting the tokenization process for a broker...");
-	console.log("1. Minting an empty batch...");
-	const batchTx = await cBatches.connect(broker).mintEmptyBatch(broker.address);
-	const batchId = await getBatchId(batchTx);
-	console.log(`Minted batch ${batchId}`);
-
-	// Confirm credit tokenization
-	console.log("2. Updating the batch with data...");
-	await cBatches.connect(admin).updateBatchWithData(batchId, "serialNumber1", 2000, "uri1");
-	console.log("3. Linking the batch to the vintage...");
-	await cBatches.connect(admin).linkWithVintage(batchId, vintageId);
-	console.log("4. Confirming the batch...");
-	await cBatches.connect(admin).confirmRetirement(batchId);
-
-	// Broker can now fractionalize to TCO2
-	console.log("5. Fractionalizing the batch to TCO2...");
-	await cBatches.connect(broker).fractionalize(batchId);
-
-	return [registry, tco2Factory, tco2];
+	let data = {};
+	for (const project of genesis) {
+		console.log(`Creating a new carbon project ${project.ProjectId}...`);
+		const pTx = await cProjects.addNewProject(
+			admin.address,
+			project.ProjectId,
+			project.Standard,
+			project.Methodology,
+			project.Region,
+			project.StorageMethod,
+			project.Method,
+			project.EmissionType,
+			project.Category,
+			project.Uri
+		);
+		const projectId = await getProjectTokenId(pTx);
+		console.log(`Created carbon project ${projectId}`);
+
+		for (const vintage of project.Vintages) {
+			console.log(`Creating a new vintage for carbon project ${projectId}...`);
+			const vTx = await cVintages.addNewVintage(admin.address, {
+				projectTokenId: projectId,
+				name: vintage.Name,
+				startTime: Date.parse(vintage.StartTime) / 1000,
+				endTime: Date.parse(vintage.EndTime) / 1000,
+				totalVintageQuantity: vintage.TotalVintageQuantity, // tonnes
+				isCorsiaCompliant: vintage.IsCorsiaCompliant,
+				isCCPcompliant: vintage.IsCCPcompliant,
+				coBenefits: vintage.CoBenefits,
+				correspAdjustment: vintage.CorrespAdjustment,
+				additionalCertification: vintage.AdditionalCertification,
+				uri: vintage.Uri,
+			});
+			const vintageId = await getProjectVintageTokenId(vTx);
+			console.log(`Created vintage ${vintageId}`);
+
+			// Deploy TCO2
+			console.log(`Deploying a TCO2 for vintage ${vintageId}...`);
+			await tco2Factory.deployFromVintage(vintageId);
+			const tco2Array = await tco2Factory.getContracts();
+			const tco2 = await ethers.getContractAt("ToucanCarbonOffsets", tco2Array[tco2Array.length - 1]);
```

Comment:
@pheuberger @kargakis In the previous code it was looking at `tco2Array[0]` ... it seems that the latest contract added is at the end of this list, but perhaps there's a better way to do this? It's weird that `deployFromVintage` doesn't return the newly created contract...

Reply:
agreed, it would've been nice if the function returned the address. anyhow, seems like a reasonable approach. you could also have a look into getting it out of an event like here: https://github.com/pheuberger/tokenizer/blob/5d50011268b0e9648871d6cfd8927092c0b88339/packages/hardhat/lib/setup-testenv.ts#L147-L151

the advantage being, you don't have to query contracts for it

Reply:
OK if you think this is a reasonable approach I'm happy to merge this (I can't merge to this repo)

Reply:
@kargakis do you agree with my assessment? if so please go ahead and merge ☺️

Reply:
In the previous version of the code, there was only a single TCO2 deployed so first and last was the same. `deployFromVintage` actually writes the [TCO2 address in the blockchain log via an event](https://github.com/ToucanProtocol/contracts/blob/cf3d9e57ab4a6712d950440e71adb917ea4f8ea2/contracts/ToucanCarbonOffsetsFactory.sol#L149), so we could create our own wrapper in JS to read the event (same way as [here](https://github.com/regen-network/toucan-bridge/blob/8b5238d9aa6c4a7682bcb732138f1cbdd14f04e4/lib/toucan.js#L15)) and return the address if we wanted to to make this more robust. For now this way is fine by me.